File: core/src/core2/vector/extensions/XtExtensionVector.java
package core2.vector.extensions;

import org.apache.arrow.memory.BufferAllocator;
import org.apache.arrow.memory.util.hash.ArrowBufHasher;
import org.apache.arrow.vector.ExtensionTypeVector;
import org.apache.arrow.vector.FieldVector;
import org.apache.arrow.vector.ValueVector;
import org.apache.arrow.vector.types.pojo.Field;
import org.apache.arrow.vector.types.pojo.FieldType;
import org.apache.arrow.vector.util.TransferPair;

public abstract class XtExtensionVector<V extends FieldVector> extends ExtensionTypeVector<V> {
    private final Field field;

    public XtExtensionVector(String name, BufferAllocator allocator, FieldType fieldType, V underlyingVector) {
        super(name, allocator, underlyingVector);
        this.field = new Field(name, fieldType, null);
    }

    public XtExtensionVector(Field field, BufferAllocator allocator, V underlyingVector) {
        super(field, allocator, underlyingVector);
        this.field = field;
    }

    @Override
    public Field getField() {
        return field;
    }

    @Override
    public int hashCode(int index) {
        return getUnderlyingVector().hashCode(index);
    }

    @Override
    public int hashCode(int index, ArrowBufHasher hasher) {
        return getUnderlyingVector().hashCode(index, hasher);
    }

    @Override
    public void copyFromSafe(int fromIndex, int thisIndex, ValueVector from) {
        getUnderlyingVector().copyFromSafe(fromIndex, thisIndex, ((XtExtensionVector<?>) from).getUnderlyingVector());
    }

    @Override
    public TransferPair makeTransferPair(ValueVector target) {
        ValueVector targetUnderlying = ((XtExtensionVector<?>) target).getUnderlyingVector();
        TransferPair tp = getUnderlyingVector().makeTransferPair(targetUnderlying);

        return new TransferPair() {
            @Override
            public void transfer() {
                tp.transfer();
            }

            @Override
            public void splitAndTransfer(int startIndex, int length) {
                tp.splitAndTransfer(startIndex, length);
            }

            @Override
            public ValueVector getTo() {
                return target;
            }

            @Override
            public void copyValueSafe(int from, int to) {
                tp.copyValueSafe(from, to);
            }
        };
    }
}

File: core/src/main/kotlin/xtdb/vector/extensions/XtExtensionVector.kt
package xtdb.vector.extensions

import org.apache.arrow.memory.BufferAllocator
import org.apache.arrow.memory.util.hash.ArrowBufHasher
import org.apache.arrow.vector.ExtensionTypeVector
import org.apache.arrow.vector.FieldVector
import org.apache.arrow.vector.ValueVector
import org.apache.arrow.vector.types.pojo.Field
import org.apache.arrow.vector.types.pojo.FieldType
import org.apache.arrow.vector.util.TransferPair

abstract class XtExtensionVector<V : FieldVector>(private val field: Field, allocator: BufferAllocator, underlyingVector: V) :
    ExtensionTypeVector<V>(field, allocator, underlyingVector) {

    constructor(name: String, allocator: BufferAllocator, fieldType: FieldType, underlyingVector: V)
        : this(Field(name, fieldType, null), allocator, underlyingVector)

    override fun getField(): Field = field

    override fun hashCode(index: Int): Int = underlyingVector.hashCode(index)
    override fun hashCode(index: Int, hasher: ArrowBufHasher): Int = underlyingVector.hashCode(index, hasher)

    override fun copyFromSafe(fromIndex: Int, thisIndex: Int, from: ValueVector) {
        underlyingVector.copyFromSafe(fromIndex, thisIndex, (from as XtExtensionVector<*>).underlyingVector)
    }

    abstract fun getObject0(index: Int) : Any

    override fun getObject(index: Int): Any? = if (isNull(index)) null else getObject0(index)

    override fun getTransferPair(field: Field, allocator: BufferAllocator) =
        makeTransferPair(field.createVector(allocator))

    override fun makeTransferPair(target: ValueVector): TransferPair {
        val targetUnderlying: ValueVector = (target as XtExtensionVector<*>).underlyingVector
        val tp = underlyingVector.makeTransferPair(targetUnderlying)

        return object : TransferPair {
            override fun transfer() = tp.transfer()
            override fun splitAndTransfer(startIndex: Int, length: Int) = tp.splitAndTransfer(startIndex, length)
            override fun getTo(): ValueVector = target
            override fun copyValueSafe(from: Int, to: Int) = tp.copyValueSafe(from, to)
        }
    }
}
